### PR TITLE
feat(routes): use named routes in core 

### LINKED ISSUE #9126: Deprecate 'handler' logic in entity menu
I am working on a different menu system, and one of the most painful things is that you can not display an entity menu in arbitrary place without knowing it's handler. Many enities rely on `handler` param to be set for `edit` and `delete` items to be registered. This is problematic if you are working on a generic solution that should work on any entity, as handlers are not predictable from entity type or subtype. Re-registering these items is also not possible because edit page location and delete action path are unpredictable.

### PR DESCRIPTION
**Pull request by hypeJunction**
Rewrites page handlers using new named route API.
Updates some resource views with logic contained in removed page handler functions.
Deprecates the use of handler parameter in entity menus and removes it
for some core entities.
Add elgg_generate_entity_url().
Add elgg_generate_action_url().

Refs #11563
Fixes #9126

### CODE REVIEW

File: mod/activity/elgg-plugin.php [hunk 1]
```diff
@@ -19,4 +19,13 @@
 			'context' => ['profile', 'dashboard'],
 		],
 	],
+	'routes' => [
+		'activity' => [
+			'path' => '/activity/{page_type}/{username?}',
```
Comment (mrclay) [inline mod/activity/elgg-plugin.php]: I think better to have multiple routes. Even though same view can handle them, different logical resources.

Comment (hypeJunction) [reply mod/activity/elgg-plugin.php]: Sure

**Comment by hypeJunction** [comment]
@mrclay Thoughts on `elgg_generate_entity_url()`? It's a bit corky and will probably fail at some point, but it is doing the job until proven otherwise.

File: engine/lib/pagehandler.php [hunk 2]
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
...
+			$type,
+			$subtype,
+			$subresource,
+		];
+
+		return implode(':', array_filter($route_parts));
+	};
+
+	$pairs = [
+		[$entity->getType(), $entity->getSubtype()],
```
Comment (mrclay) [inline engine/lib/pagehandler.php]: `->type`, `->subtype` thanks to you.

File: engine/lib/pagehandler.php [hunk 3]
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
...
+		}
+
+		switch ($prop) {
+			case 'title' :
+			case 'name' :
+				$query[$prop] = elgg_get_friendly_title($entity->getDisplayName());
+				break;
+
+			default :
+				$query[$prop] = $entity->$prop;
```
Comment (mrclay) [inline engine/lib/pagehandler.php]: Use case for this?

File: engine/lib/pagehandler.php [hunk 4]
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
...
 }
 
+/**
+ * Generate entity URL from a named route
+ * Route segments will be automatically resolved from entity attributes and metadata
+ *
+ * @param ElggEntity $entity      Entity
+ * @param string     $resource    Resource name
+ * @param string     $subresource Subresource name
+ * @param array      $query       URL query elements
```
Comment (mrclay) [inline engine/lib/pagehandler.php]: Rather this be $params. $query makes me think querystring.

File: engine/lib/pagehandler.php [hunk 5]
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
...
+ *
+ * @param string $action          Action name
+ * @param array  $query           Query elements
+ * @param bool   $add_csrf_tokens Add tokens
+ *
+ * @return string
+ */
+function elgg_generate_action_url($action, array $query = [], $add_csrf_tokens = true) {
+
+	$url = "action/$action";
```
Comment (mrclay) [inline engine/lib/pagehandler.php]: Yes! Never wondering if it's "actions/" or "action/".

File: engine/lib/pagehandler.php (hunk 3 above)
Comment (hypeJunction) [reply engine/lib/pagehandler.php]: `guid`, `container_guid`, `owner_guid` etc.
I imagine someone using `time_created` or some other metadata in the route

File: engine/lib/pagehandler.php (hunk 4 above)
Comment (hypeJunction) [reply engine/lib/pagehandler.php]: Anything that's not defined as a URL segment goes into query string

Comment (hypeJunction) [reply engine/lib/pagehandler.php]: I will rename and update the docs for both

File: engine/lib/pagehandler.php (hunk 5 above)
Comment (hypeJunction) [reply engine/lib/pagehandler.php]: Not to mention `"action/friends/add?guid=" . elgg_get_page_owner_guid()`

File: engine/lib/navigation.php [hunk 6]
```diff
@@ -627,39 +627,47 @@ function _elgg_page_menu_setup(\Elgg\Hook $hook) {
...
+		elgg_deprecated_notice("Using 'handler' in entity menu parameters is deprecated. Use named routes instead.", '3.0');
+
+		$edit_url = "$handler/edit/{$entity->guid}";
+
+		if (elgg_action_exists("$handler/delete")) {
+			$action = "$handler/delete";
+			$delete_url = elgg_generate_action_url($action, [
+				'guid' => $entity->guid,
+			]);
+		}
```
Comment (jdalsem) [inline engine/lib/navigation.php]: should fallback to `entity/delete`

**Comment by jdalsem** [comment]
needs rebase

**Comment by hypeJunction** [comment]
FFS

**Pull request merged**